gromacs/tools.py

```diff
@@ -94,177 +108,262 @@ def _generate_sphinx_class_string(clsname):
+    """
+    drivers = config.get_tool_names()
+
+    if len(drivers) == 0 and 'GMXBIN' in os.environ:
+        drivers = find_executables(os.environ['GMXBIN'])
+
+    if len(drivers) == 0 or len(drivers) > 4:
+        drivers = ['gmx', 'gmx_d', 'gmx_mpi', 'gmx_mpi_d']
+
+    tools = {}
+    for driver in drivers:
+        try:
+            out = subprocess.check_output([driver, '-quiet', 'help',
+                                           'commands'])
+            for line in str(out).encode('ascii').splitlines()[5:-1]:
+                if line[4] != ' ':
+
+                    name = line[4:line.index(' ', 4)]
+                    fancy = make_valid_identifier(name)
+                    suffix = driver.partition('_')[2]
+                    if suffix:
+                        fancy = '%s_%s' % (fancy, suffix)
+                    tools[fancy] = tool_factory(fancy, name, driver)
+        except (subprocess.CalledProcessError, OSError):
+            pass
+
+    if len(tools) == 0:
+        raise GromacsToolLoadingError("Failed to load v5 tools")
+    return tools
+
+
+def load_v4_tools():
+    """ Load Gromacs 4.x tools automatically using some heuristic.
+
+    Tries to load tools (1) in configured tool groups (2) and fails back  to
+    automatic detection from ``GMXBIN`` (3) then to a prefilled list.
+
+    :return: dict mapping tool names to GromacsCommand classes
+    """
+    names = config.get_tool_names()
+
+    if len(names) == 0 and 'GMXBIN' in os.environ:
+        names = find_executables(os.environ['GMXBIN'])
+
+    if len(names) == 0 or len(names) > len(V4TOOLS) * 4:
+        names = V4TOOLS[:]
+
+    tools = {}
+    for name in names:
+        try:
+            null = open(os.devnull, 'w')
+            subprocess.check_call([name], stdout=null, stderr=null)
+        except subprocess.CalledProcessError:
+            pass
+        fancy = make_valid_identifier(name)
+        tools[fancy] = tool_factory(fancy, name, None)
+
+    if len(tools) == 0:
+        raise GromacsToolLoadingError("Failed to load v4 tools")
+    return tools
+
+
+def load_extra_tools():
+    """ Load extra tools.
+
+    :return: dict mapping tool names to GromacsCommand classes
+    """
+    names = config.cfg.get("Gromacs", "extra").split()
+    tools = {}
+    for name in names:
+        fancy = make_valid_identifier(name)
+        driver = 'gmx' if config.MAJOR_RELEASE == '5' else None
+        tools[name] = tool_factory(fancy, name, driver)
+    return tools
+
+
+def merge_ndx(*args):
+    """ Takes one or more index files and optionally one structure file and
+    returns a path for a new merged index file.
+
+    :param args: index files and zero or one structure file
+    :return: path for the new merged index file
+    """
+    ndxs = []
+    struct = None
+    for fname in args:
+        if fname.endswith('.ndx'):
+            ndxs.append(fname)
+        else:
+            assert struct is None, "only one structure file supported"
```

Comment:
Don't use `assert` to check user input – one cannot easily work with it in user code and when run with `python -O`, asseryions are optimized away. Just raise an exception with `ValueError`.